File: OneToolkit/FileSizeHelper.cpp
#include "pch.h"
#include "FileSizeHelper.h"
#include "Storage.FileSizeHelper.g.cpp"

using namespace juv;

namespace winrt::OneToolkit::Storage::implementation
{
	constexpr std::array BinaryUnits{ L"B", L"KiB", L"MiB", L"GiB", L"TiB", L"PiB", L"EiB", L"ZiB", L"YiB" };
	constexpr std::array DecimalUnits{ L"B", L"KB", L"MB", L"GB", L"TB", L"PB", L"EB", L"ZB", L"YB" };

	hstring FileSizeHelper::ToFormattedString(uint64 size, bool useBinaryPrefix)
	{
		uint8 index = 0;
		auto preciseSize = static_cast<double>(size);
		const auto multiple = useBinaryPrefix ? 1024 : 1000;
		while (preciseSize >= multiple)
		{
			preciseSize /= multiple;
			++index;
		}

		if (index > 8)
		{
			index = 0;
			preciseSize = static_cast<double>(size);
		}

		return hstring(std::format(L"{0} {1}", preciseSize, useBinaryPrefix ? BinaryUnits[index] : DecimalUnits[index]));
	}

	uint64 FileSizeHelper::FromFormattedString(hstring const& formattedString, bool useBinaryPrefix)
	{
		throw hresult_not_implemented();
	}
}

File: OneToolkit.Dotnet/Storage/FileSizeHelper.cs
﻿using System;

namespace OneToolkit.Storage
{
	/// <summary>
	/// Provides static helper methods to convert file size to a human readable string and vice versa.
	/// </summary>
	public static class FileSizeHelper
	{
		private static string[] BinaryUnits = { "B", "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB" };

		private static string[] DecimalUnits = { "B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB" };

		/// <summary>
		/// Formats file size to a human readable string.
		/// </summary>
		public static string Stringify(ulong sizeInBytes, bool useBinaryPrefix)
		{
			byte index = 0;
			var preciseSize = Convert.ToDouble(sizeInBytes);
			var multiple = useBinaryPrefix ? 1024 : 1000;
			while (preciseSize >= multiple)
			{
				preciseSize /= multiple;
				++index;
			}

			if (index > 8)
			{
				index = 0;
				preciseSize = Convert.ToDouble(sizeInBytes);
			}

			return $"{preciseSize} {(useBinaryPrefix ? BinaryUnits[index] : DecimalUnits[index])}";
		}

		/// <summary>
		/// Tries to convert a human readable file size string to an integer representing the size in bytes.
		/// </summary>
		public static ulong Parse(string formattedString)
		{
			if (formattedString.EndsWith(" B")) return Convert.ToUInt64(formattedString.Substring(0, formattedString.Length - 2));
			throw new NotImplementedException();
		}
	}
}